### ISSUE TITLE
Joblib hangs(Randomly) after finishes the parallel for loop

### ISSUE THREAD

**Issue by simonzcaiman** [joblib/joblib#502]
Hi guys, 

I'm experience hangs after the Parallel for loop finished. I run my code with verbose = 200 so that I can see the std out. 
Below is the last line of console output before it hangs. 
"[Parallel(n_jobs=12)]: Done 101 out of 100 | elapsed:  1.7min remaining:   -1.0s"

In contrast, here the std out when it did not hang
"[Parallel(n_jobs=12)]: Done 101 out of 100 | elapsed:  1.7min remaining:   -1.0s
[Parallel(n_jobs=12)]: Done 100 out of 100 | elapsed:  1.7min finished
deling <__main__.Genetics object at 0x7f9bcdc47ef0>
deling <__main__.Genetics object at 0x7f9bcdc47f60>
deling <__main__.Genetics object at 0x7f9bceff0f98>
deling <__main__.Genetics object at 0x7f9bcdc35208>
deling <__main__.Genetics object at 0x7f9bcee5e860>
deling <__main__.Genetics object at 0x7f9bcee5edd8>
deling <__main__.Genetics object at 0x7f9bcdc35240>
deling <__main__.Genetics object at 0x7f9bcdc24908>
deling <__main__.Genetics object at 0x7f9bcdc47128>
deling <__main__.Genetics object at 0x7f9bcdc35208>
deling <__main__.Genetics object at 0x7f9bcdc24048>
deling <__main__.Genetics object at 0x7f9bceff0438>"

The function I'm passing to the Parallel object is a member of an object. The augment that I'm passing in the delayed function is an self defined object, which has pandas dataframe in it. 
Most of the time, the program run just fine without hanging. 
In addition, I see " Done 101 out of 100 "being printed here. Is this normal? Am I doing something wrong here to cause this printing? 

I have a for loop for 100 iterations. Why does this keep saying Done 101 out of 100? 
Any help could be appreciated. Thanks in advance.

**Comment by simonzcaiman** [comment]
In addition, memory is at a 20% level. There is no problem with the function itself.

**Comment by lesteve** [comment]
> "[Parallel(n_jobs=12)]: Done 101 out of 100 | elapsed: 1.7min remaining: -1.0s"

Which version of joblib are you using? The negative time remaining was fixed a little while ago. You probably want to try with the latest released joblib.

**Comment by simonzcaiman** [comment]
Thanks for the reply. I was using 0.9.4 version, which I installed using "conda install joblib". I uninstall it and install the newest version using pip install, which gave me the lastest version. 
Now the output is like this: 
[Parallel(n_jobs=12)]: Done 4998 out of 5000 | elapsed: 69.0min remaining:    1.7s
"[Parallel(n_jobs=12)]: Done 5000 out of 5000 | elapsed: 69.0min finished"
I run three instances last time with 5K and 10K iterations for loop. It worked just fine. No freeze anymore. 
I believe the new version helped. Thanks for the hard work putting in this library.

**Comment by lesteve** [comment]
Great, thanks for following up on this! Don't hesitate to reopen the issue if you notice this problem again.

**Comment by lesteve** [comment]
I see that you have reopened the issue, which I guess means that you have seen the problem again. It would be great to post a stand-alone snippet so we can try to reproduce the issue.

Also for future reference can you post the output of:
```py
import platform; print(platform.platform())
import sys; print("Python", sys.version)
import joblib; print("joblib", joblib.__version__)
```

**Comment by simonzcaiman** [comment]
Hi , 
The same problem occurred to me again last night when I was trying run a parallel for loop for 5000 iterations. I ran the same script on three different servers. 
All of them got struck at the second last iterations. Here is the print out when I set verbose 50+
The last print out before it hangs:
"[Parallel(n_jobs=12)]: Done 4998 out of 5000 | elapsed: 11.3min remaining:    0.3s"
I believe it finished running the 4998th job. 
Since I have some print out for each run, 4999th thread was able to finish its job. However, it hung there and no more print out after that.  

Here is my parallel for loop code:

num_cores = multiprocessing.cpu_count()
graded = Parallel(n_jobs=num_cores,verbose=250)(delayed(self.fitness)(x,TrainingSetStartTime,TrainingSetEndTime) for x in pop)


self.fitness is a function that will consume a large pandas dataframe, turn it into a numpy array ,do some extensive calculations and create a pandas dataframe object at the end and return it. 

Is that possible that the large pandas dataframe object or numpy array that is causing the problem? 

I don't see any bug/exception being raised here. The program doesn't always freeze. However when it freezes, it always freezes at the end of its second last job. 
I updated the joblib up to 0.10.3. 
Is there anything else I can provide to better explain the issue here? 
Thanks a lot for your help and support!

**Comment by simonzcaiman** [comment]
Here is the output: 
Linux-3.10.0-229.el7.x86_64-x86_64-with-centos-7.1.1503-Core
Python 3.5.2 |Anaconda 4.0.0 (64-bit)| (default, Jul  2 2016, 17:53:06) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]
joblib 0.10.3

**Comment by simonzcaiman** [comment]
Alright, I will try my best to reproduce the same issue with a stand-alone snippet. Meanwhile, if you see anything suspicious from my problem description above, please let me know. thanks!

**Comment by lesteve** [comment]
I just released joblib 0.11, can you try installing it and seeing if the problem persists? I don't think that it will fix your problem but at least you will use the latest released code.

**Comment by simonzcaiman** [comment]
Great! Thank you. I will give it a try. thanks!

**Comment by simonzcaiman** [comment]
Hi,
   Just updates from me. Apparently, in the case when the loop freezes, it finished task in the last task, but it never went to print out the process saying "[Parallel(n_jobs=12)]: Done 300 out of 300 | elapsed:   46.6s finished "

 
Here is the stdout from the program without freezing: 

[Parallel(n_jobs=12)]: Done 295 out of 300 | elapsed:   46.0s remaining:    0.8s
[Parallel(n_jobs=12)]: Done 297 out of 300 | elapsed:   46.4s remaining:    0.5s
[Parallel(n_jobs=12)]: Done 300 out of 300 | elapsed:   46.6s finished

Here is the stdout from the program with freezing: 

[Parallel(n_jobs=12)]: Done 295 out of 300 | elapsed:   46.0s remaining:    0.8s
[Parallel(n_jobs=12)]: Done 297 out of 300 | elapsed:   46.4s remaining:    0.5s


Is it possible that the something happened after the last thread finished its assigned task? which caused the last thread not be able to release the lock(which will prevent the main thread from continue running) ?

**Comment by simonzcaiman** [comment]
Hey, 
Another observation I have is that After the original job iterator becomes None( fully consumed), I am only getting print progress once per two tasks completed. 
I have 300 jobs with 12 cores(n_jobs)

[Parallel(n_jobs=12)]: Done 275 tasks      | elapsed:   43.1s
[Parallel(n_jobs=12)]: Done 276 tasks      | elapsed:   43.2s
[Parallel(n_jobs=12)]: Done 277 tasks      | elapsed:   43.4s
[Parallel(n_jobs=12)]: Done 279 out of 300 | elapsed:   43.8s remaining:    3.3s
[Parallel(n_jobs=12)]: Done 281 out of 300 | elapsed:   44.0s remaining:    3.0s
[Parallel(n_jobs=12)]: Done 283 out of 300 | elapsed:   44.2s remaining:    2.7s
[Parallel(n_jobs=12)]: Done 285 out of 300 | elapsed:   44.7s remaining:    2.4s
[Parallel(n_jobs=12)]: Done 287 out of 300 | elapsed:   44.8s remaining:    2.0s
[Parallel(n_jobs=12)]: Done 289 out of 300 | elapsed:   45.2s remaining:    1.7s
[Parallel(n_jobs=12)]: Done 291 out of 300 | elapsed:   45.5s remaining:    1.4s
[Parallel(n_jobs=12)]: Done 293 out of 300 | elapsed:   45.8s remaining:    1.1s
[Parallel(n_jobs=12)]: Done 295 out of 300 | elapsed:   46.0s remaining:    0.8s
[Parallel(n_jobs=12)]: Done 297 out of 300 | elapsed:   46.4s remaining:    0.5s

And then without freezing it will produce another line:
[Parallel(n_jobs=12)]: Done 300 out of 300 | elapsed:   46.2s finished

Otherwise, it will just freeze before printing out the above line.

**Comment by simonzcaiman** [comment]
Is it possible that some sort of racing condition (which caused the randomness in this case) mess up the join/async operation for the main thread to continue?  The wired thing is the freezing is always happening right at the end of the parallel for loop completion.

**Comment by simonzcaiman** [comment]
Hey, 
In the ApplyResult class, get function, if self._success not true, why is it raising self._value? 
If it raise self._value, the catch function is not able to pick up since it is not an exception. I'm thinking this might be the problem.

**Comment by simonzcaiman** [comment]
I managed to track the error down to this line of code. 
In ApplyResult, get function, "self.wait(timeout)"
If ApplyResult._event.wait(timeout) never got notified, the program will get struct there,which will ultimately caused the freeze. I think in my case, the event object 's internal flag never turned to true.

**Comment by simonzcaiman** [comment]
Ok. I got it. If there is timeout=None and the thread got struck when executing the function, the entire parallel for loop will freeze. Setting the timeout to some reasonable float will help to identify the suspicious task with error.

**Comment by simonzcaiman** [comment]
I think you might want to consider putting on an option saying that one thread exceeded the time out number, return a default value to return, so that the parallel for loop can keep running.

**Comment by GaelVaroquaux** [comment]
> Closed #502.

Why did you close? You seemed like you were onto something.
